src/libstore/outputs-spec.hh

```diff
@@ -1,32 +1,91 @@
 #pragma once
 
+#include <optional>
+#include <set>
 #include <variant>
 
-#include "util.hh"
-
-#include "nlohmann/json_fwd.hpp"
+#include "json-impls.hh"
 
 namespace nix {
 
-typedef std::set<std::string> OutputNames;
+struct OutputNames : std::set<std::string> {
+    using std::set<std::string>::set;
+
+    // These need to be "inherited manually"
+    OutputNames(const std::set<std::string> & s)
+        : std::set<std::string>(s)
+    { }
+    OutputNames(std::set<std::string> && s)
+        : std::set<std::string>(s)
+    { }
 
-struct AllOutputs {
-    bool operator < (const AllOutputs & _) const { return false; }
+    /* This set should always be non-empty, so we delete this
+       constructor in order make creating empty ones by mistake harder.
+       */
+    OutputNames() = delete;
 };
 
-struct DefaultOutputs {
-    bool operator < (const DefaultOutputs & _) const { return false; }
+struct AllOutputs : std::monostate { };
+
+typedef std::variant<AllOutputs, OutputNames> _OutputsSpecRaw;
+
+struct OutputsSpec : _OutputsSpecRaw {
```

Comment:
And here we have another subclass of STL standard classes... please, please, don't.

Reply:
See https://github.com/NixOS/nix/issues/7479 we have an issue discussing this pattern. I am for it as the best that can be done for Sum types until we switch to Rust.